File: benchmarks/JavaScript/richards.js
// @ts-check
// The benchmark in its current state is a derivation from the SOM version,
// which is derived from Mario Wolczko's Smalltalk version of DeltaBlue.
//
// The original license details are availble here:
// http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html

const { Benchmark } = require('./benchmark');

const NO_TASK = null;
const NO_WORK = null;
const IDLER = 0;
const WORKER = 1;
const HANDLER_A = 2;
const HANDLER_B = 3;
const DEVICE_A = 4;
const DEVICE_B = 5;
const NUM_TYPES = 6;

const DEVICE_PACKET_KIND = 0;
const WORK_PACKET_KIND = 1;

const DATA_SIZE = 4;

const TRACING = false;

class RBObject {
  append(packet, queueHead) {
    packet.link = NO_WORK;
    if (NO_WORK === queueHead) {
      return packet;
    }

    let mouse = queueHead;
    let link;

    while (NO_WORK !== (link = mouse.link)) {
      mouse = link;
    }
    mouse.link = packet;
    return queueHead;
  }
}

class TaskState extends RBObject {
  constructor() {
    super();
    this.packetPending_ = false;
    this.taskWaiting_ = false;
    this.taskHolding_ = false;
  }

  isPacketPending() { return this.packetPending_; }

  isTaskHolding() { return this.taskHolding_; }

  isTaskWaiting() { return this.taskWaiting_; }

  setTaskHolding(b) { this.taskHolding_ = b; }

  setTaskWaiting(b) { this.taskWaiting_ = b; }

  setPacketPending(b) { this.packetPending_ = b; }

  packetPending() {
    this.packetPending_ = true;
    this.taskWaiting_ = false;
    this.taskHolding_ = false;
  }

  running() {
    this.packetPending_ = false;
    this.taskWaiting_ = false;
    this.taskHolding_ = false;
  }

  waiting() {
    this.packetPending_ = false;
    this.taskHolding_ = false;
    this.taskWaiting_ = true;
  }

  waitingWithPacket() {
    this.taskHolding_ = false;
    this.taskWaiting_ = true;
    this.packetPending_ = true;
  }

  isTaskHoldingOrWaiting() {
    return this.taskHolding_ || (!this.packetPending_ && this.taskWaiting_);
  }

  isWaitingWithPacket() {
    return this.packetPending_ && this.taskWaiting_ && !this.taskHolding_;
  }

  static createRunning() {
    const t = new TaskState();
    t.running();
    return t;
  }

  static createWaiting() {
    const t = new TaskState();
    t.waiting();
    return t;
  }

  static createWaitingWithPacket() {
    const t = new TaskState();
    t.waitingWithPacket();
    return t;
  }
}

class DeviceTaskDataRecord extends RBObject {
  constructor() {
    super();
    this.pending = NO_WORK;
  }
}

class HandlerTaskDataRecord extends RBObject {
  constructor() {
    super();
    this.workIn = NO_WORK;
    this.deviceIn = NO_WORK;
  }

  deviceInAdd(packet) {
    this.deviceIn = this.append(packet, this.deviceIn);
  }

  workInAdd(packet) {
    this.workIn = this.append(packet, this.workIn);
  }
}

class IdleTaskDataRecord extends RBObject {
  constructor() {
    super();
    this.control = 1;
    this.count = 10000;
  }
}

class Packet extends RBObject {
  constructor(link, identity, kind) {
    super();
    this.link = link;
    this.identity = identity;
    this.kind = kind;
    this.datum = 0;
    this.data = new Array(DATA_SIZE).fill(0);
  }
}

class TaskControlBlock extends TaskState {
  constructor(link, identity, priority, initialWorkQueue, initialState, privateData, fn) {
    super();
    this.link = link;
    this.identity = identity;
    this.priority = priority;
    this.input = initialWorkQueue;
    this.setPacketPending(initialState.isPacketPending());
    this.setTaskWaiting(initialState.isTaskWaiting());
    this.setTaskHolding(initialState.isTaskHolding());
    this.handle = privateData;
    this.function = fn;
  }

  addInputAndCheckPriority(packet, oldTask) {
    if (NO_WORK === this.input) {
      this.input = packet;
      this.setPacketPending(true);
      if (this.priority > oldTask.priority) { return this; }
    } else {
      this.input = this.append(packet, this.input);
    }
    return oldTask;
  }

  runTask() {
    let message;
    if (this.isWaitingWithPacket()) {
      message = this.input;
      this.input = message.link;
      if (NO_WORK === this.input) {
        this.running();
      } else {
        this.packetPending();
      }
    } else {
      message = NO_WORK;
    }
    return this.function(message, this.handle);
  }
}

class WorkerTaskDataRecord extends RBObject {
  constructor() {
    super();
    this.destination = HANDLER_A;
    this.count = 0;
  }
}

class Scheduler extends RBObject {
  constructor() {
    super();

    // init tracing
    this.layout = 0;

    // init scheduler
    this.queuePacketCount = 0;
    this.holdCount = 0;
    this.taskTable = new Array(NUM_TYPES).fill(NO_TASK);
    this.taskList = NO_TASK;

    this.currentTask = null;
    this.currentTaskIdentity = 0;
  }

  createDevice(identity, priority, workPacket, state) {
    const data = new DeviceTaskDataRecord();

    this.createTask(identity, priority, workPacket, state, data, (workArg, wordArg) => {
      const dataRecord = wordArg;
      let functionWork = workArg;
      if (NO_WORK === functionWork) {
        if (NO_WORK === (functionWork = dataRecord.pending)) {
          return this.markWaiting();
        }
        dataRecord.pending = NO_WORK;
        return this.queuePacket(functionWork);
      }
      dataRecord.pending = functionWork;
      if (TRACING) {
        this.trace(functionWork.datum);
      }
      return this.holdSelf();
    });
  }

  createHandler(identity, priority, workPacket, state) {
    const data = new HandlerTaskDataRecord();
    this.createTask(identity, priority, workPacket, state, data, (work, word) => {
      const dataRecord = word;
      if (NO_WORK !== work) {
        if (WORK_PACKET_KIND === work.kind) {
          dataRecord.workInAdd(work);
        } else {
          dataRecord.deviceInAdd(work);
        }
      }

      let workPacket_;
      if (NO_WORK === (workPacket_ = dataRecord.workIn)) {
        return this.markWaiting();
      }
      const count = workPacket_.datum;
      if (count >= DATA_SIZE) {
        dataRecord.workIn = workPacket_.link;
        return this.queuePacket(workPacket_);
      }
      let devicePacket;
      if (NO_WORK === (devicePacket = dataRecord.deviceIn)) {
        return this.markWaiting();
      }
      dataRecord.deviceIn = devicePacket.link;
      devicePacket.datum = workPacket_.data[count];
      workPacket_.datum = count + 1;
      return this.queuePacket(devicePacket);
    });
  }

  createIdler(identity, priority, work, state) {
    const data = new IdleTaskDataRecord();
    this.createTask(identity, priority, work, state, data, (workArg, wordArg) => {
      const dataRecord = wordArg;
      dataRecord.count -= 1;
      if (0 === dataRecord.count) {
        return this.holdSelf();
      }
      if (0 === (dataRecord.control & 1)) {
        dataRecord.control /= 2;
        return this.release(DEVICE_A);
      }
      dataRecord.control = (dataRecord.control / 2) ^ 53256;
      return this.release(DEVICE_B);
    });
  }

  createPacket(link, identity, kind) {
    return new Packet(link, identity, kind);
  }

  createTask(identity, priority, work, state, data, fn) {
    const t = new TaskControlBlock(
      this.taskList,
      identity,
      priority,
      work,
      state,
      data,
      fn
    );
    this.taskList = t;
    this.taskTable[identity] = t;
  }

  createWorker(identity, priority, workPacket, state) {
    const dataRecord = new WorkerTaskDataRecord();

    this.createTask(
      identity,
      priority,
      workPacket,
      state,
      dataRecord,
      (work, word) => {
        const data = word;
        if (NO_WORK === work) {
          return this.markWaiting();
        }

        data.destination = (HANDLER_A === data.destination) ? HANDLER_B : HANDLER_A;
        work.identity = data.destination;
        work.datum = 0;
        for (let i = 0; i < DATA_SIZE; i += 1) {
          data.count += 1;
          if (data.count > 26) { data.count = 1; }
          work.data[i] = 65 + data.count - 1;
        }
        return this.queuePacket(work);
      }
    );
  }

  start() {
    let workQ;

    this.createIdler(IDLER, 0, NO_WORK, TaskState.createRunning());
    workQ = this.createPacket(NO_WORK, WORKER, WORK_PACKET_KIND);
    workQ = this.createPacket(workQ, WORKER, WORK_PACKET_KIND);

    this.createWorker(WORKER, 1000, workQ, TaskState.createWaitingWithPacket());
    workQ = this.createPacket(NO_WORK, DEVICE_A, DEVICE_PACKET_KIND);
    workQ = this.createPacket(workQ, DEVICE_A, DEVICE_PACKET_KIND);
    workQ = this.createPacket(workQ, DEVICE_A, DEVICE_PACKET_KIND);

    this.createHandler(HANDLER_A, 2000, workQ, TaskState.createWaitingWithPacket());
    workQ = this.createPacket(NO_WORK, DEVICE_B, DEVICE_PACKET_KIND);
    workQ = this.createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);
    workQ = this.createPacket(workQ, DEVICE_B, DEVICE_PACKET_KIND);

    this.createHandler(HANDLER_B, 3000, workQ, TaskState.createWaitingWithPacket());
    this.createDevice(DEVICE_A, 4000, NO_WORK, TaskState.createWaiting());
    this.createDevice(DEVICE_B, 5000, NO_WORK, TaskState.createWaiting());

    this.schedule();

    return this.queuePacketCount === 23246 && this.holdCount === 9297;
  }

  findTask(identity) {
    const t = this.taskTable[identity];
    if (NO_TASK === t) { throw new Error('findTask failed'); }
    return t;
  }

  holdSelf() {
    this.holdCount += 1;
    this.currentTask.setTaskHolding(true);
    return this.currentTask.link;
  }

  queuePacket(packet) {
    const t = this.findTask(packet.identity);
    if (NO_TASK === t) { return NO_TASK; }

    this.queuePacketCount += 1;

    packet.link = NO_WORK;
    packet.identity = this.currentTaskIdentity;
    return t.addInputAndCheckPriority(packet, this.currentTask);
  }

  release(identity) {
    const t = this.findTask(identity);
    if (NO_TASK === t) { return NO_TASK; }
    t.setTaskHolding(false);
    if (t.priority > this.currentTask.priority) {
      return t;
    }
    return this.currentTask;
  }

  trace(id) {
    this.layout -= 1;
    if (0 >= this.layout) {
      process.stdout.write('\n');
      this.layout = 50;
    }
    process.stdout.write(id);
  }

  markWaiting() {
    this.currentTask.setTaskWaiting(true);
    return this.currentTask;
  }

  schedule() {
    this.currentTask = this.taskList;
    while (NO_TASK !== this.currentTask) {
      if (this.currentTask.isTaskHoldingOrWaiting()) {
        this.currentTask = this.currentTask.link;
      } else {
        this.currentTaskIdentity = this.currentTask.identity;
        if (TRACING) { this.trace(this.currentTaskIdentity); }
        this.currentTask = this.currentTask.runTask();
      }
    }
  }
}

class Richards extends Benchmark {
  benchmark() {
    return (new Scheduler()).start();
  }

  verifyResult(result) {
    return result;
  }
}

exports.newInstance = () => new Richards();

File: benchmarks/Python/richards.py
# The benchmark in its current state is a derivation from the SOM version,
# which is derived from Mario Wolczko's Smalltalk version of DeltaBlue.
#
# The original license details are availble here:
# http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html

# This file itself, and its souce control history is however based on the
# following. It is unclear whether this still bears any relevance since the
# nature of the code was essentially reverted back to the Smalltalk version.
#
# Derived from http://pws.prserv.net/dlissett/ben/bench1.htm
# Licensed CC BY-NC-SA 1.0
from benchmark import Benchmark

_NO_TASK = None
_NO_WORK = None

_IDLER = 0
_WORKER = 1
_HANDLER_A = 2
_HANDLER_B = 3
_DEVICE_A = 4
_DEVICE_B = 5

_NUM_TYPES = 6

_DEVICE_PACKET_KIND = 0
_WORK_PACKET_KIND = 1

_DATA_SIZE = 4

_TRACING = False


class Richards(Benchmark):
    def benchmark(self):
        return _Scheduler().start()

    def verify_result(self, result):
        return result


class _RBObject:
    @staticmethod
    def append(packet, queue_head):
        packet.link = _NO_WORK
        if _NO_WORK == queue_head:
            return packet

        mouse = queue_head

        while True:
            link = mouse.link
            if link == _NO_WORK:
                break
            mouse = link

        mouse.link = packet
        return queue_head


class _Scheduler(_RBObject):
    def __init__(self):
        # init tracing
        self._layout = 0

        # init scheduler
        self._task_list = _NO_TASK
        self._current_task = _NO_TASK
        self._current_task_identity = 0

        self._task_table = [_NO_TASK] * _NUM_TYPES

        self._queue_count = 0
        self._hold_count = 0

    def create_device(self, identity, priority, work, state):
        data = _DeviceTaskDataRecord()

        def fn(function_work, data_record):
            if _NO_WORK == function_work:
                function_work = data_record.pending
                if _NO_WORK == function_work:
                    return self.mark_waiting()

                data_record.pending = _NO_WORK
                return self.queue_packet(function_work)

            data_record.pending = function_work
            if _TRACING:
                self._trace(function_work.datum)
            return self._hold_self()

        self.create_task(identity, priority, work, state, data, fn)

    def create_handler(self, identity, priority, work, state):
        data = _HandlerTaskDataRecord()

        def fn(work_arg, data_record):
            if _NO_WORK != work_arg:
                if _WORK_PACKET_KIND == work_arg.kind:
                    data_record.work_in_add(work_arg)
                else:
                    data_record.device_in_add(work_arg)

            work_packet = data_record.work_in
            if _NO_WORK == work_packet:
                return self.mark_waiting()

            count = work_packet.datum
            if count >= _DATA_SIZE:
                data_record.work_in = work_packet.link
                return self.queue_packet(work_packet)

            device_packet = data_record.device_in
            if _NO_WORK == device_packet:
                return self.mark_waiting()

            data_record.device_in = device_packet.link
            device_packet.datum = work_packet.data[count]
            work_packet.datum = count + 1
            return self.queue_packet(device_packet)

        self.create_task(identity, priority, work, state, data, fn)

    def create_idler(self, identity, priority, work, state):
        data = _IdleTaskDataRecord()

        def fn(_work, data_record):
            data_record.count = data_record.count - 1
            if 0 == data_record.count:
                return self._hold_self()

            if 0 == (data_record.control & 1):
                data_record.control = data_record.control // 2
                return self.release(_DEVICE_A)

            data_record.control = (data_record.control // 2) ^ 53_256
            return self.release(_DEVICE_B)

        self.create_task(identity, priority, work, state, data, fn)

    def create_packet(self, link, identity, kind):
        return _Packet(link, identity, kind)

    def create_task(self, identity, priority, work, state, data, fn):
        t = _TaskControlBlock(
            self._task_list, identity, priority, work, state, data, fn
        )
        self._task_list = t
        self._task_table[identity] = t

    def create_worker(self, identity, priority, work, state):
        data = _WorkerTaskDataRecord()

        def fn(work, data_record):
            if _NO_WORK == work:
                return self.mark_waiting()

            data_record.destination = (
                _HANDLER_B if _HANDLER_A == data_record.destination else _HANDLER_A
            )

            work.identity = data_record.destination
            work.datum = 0
            for i in range(_DATA_SIZE):
                data_record.count = data_record.count + 1
                if data_record.count > 26:
                    data_record.count = 1
                work.data[i] = 65 + data_record.count - 1

            return self.queue_packet(work)

        self.create_task(identity, priority, work, state, data, fn)

    def start(self):
        self.create_idler(_IDLER, 0, _NO_WORK, _TaskState.with_running())
        wkq = self.create_packet(_NO_WORK, _WORKER, _WORK_PACKET_KIND)
        wkq = self.create_packet(wkq, _WORKER, _WORK_PACKET_KIND)

        self.create_worker(_WORKER, 1000, wkq, _TaskState.with_waiting_with_packet())
        wkq = self.create_packet(_NO_WORK, _DEVICE_A, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_A, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_A, _DEVICE_PACKET_KIND)

        self.create_handler(
            _HANDLER_A, 2000, wkq, _TaskState.with_waiting_with_packet()
        )
        wkq = self.create_packet(_NO_WORK, _DEVICE_B, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_B, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_B, _DEVICE_PACKET_KIND)

        self.create_handler(
            _HANDLER_B, 3000, wkq, _TaskState.with_waiting_with_packet()
        )
        self.create_device(_DEVICE_A, 4000, _NO_WORK, _TaskState.with_waiting())
        self.create_device(_DEVICE_B, 5000, _NO_WORK, _TaskState.with_waiting())

        self.schedule()

        return self._queue_count == 23_246 and self._hold_count == 9297

    def find_task(self, identity):
        t = self._task_table[identity]
        if _NO_TASK == t:
            raise Exception("find_task failed")
        return t

    def _hold_self(self):
        self._hold_count += 1
        self._current_task.set_task_holding(True)
        return self._current_task.link

    def queue_packet(self, packet):
        task = self.find_task(packet.identity)
        if _NO_TASK == task:
            return _NO_TASK

        self._queue_count += 1

        packet.link = _NO_WORK
        packet.identity = self._current_task_identity
        return task.add_input_and_check_priority(packet, self._current_task)

    def release(self, identity):
        task = self.find_task(identity)
        if _NO_TASK == task:
            return _NO_TASK

        task.set_task_holding(False)

        if task.priority > self._current_task.priority:
            return task
        return self._current_task

    def _trace(self, msg):
        self._layout -= 1
        if 0 >= self._layout:
            print("")
            self._layout = 50

        print(msg)

    def mark_waiting(self):
        self._current_task.set_task_waiting(True)
        return self._current_task

    def schedule(self):
        self._current_task = self._task_list
        while _NO_TASK is not self._current_task:
            if self._current_task.is_task_holding_or_waiting():
                self._current_task = self._current_task.link
            else:
                self._current_task_identity = self._current_task.identity
                if _TRACING:
                    self._trace(self._current_task_identity)
                self._current_task = self._current_task.run_task()


class _DeviceTaskDataRecord(_RBObject):
    def __init__(self):
        self.pending = _NO_WORK


class _HandlerTaskDataRecord(_RBObject):
    def __init__(self):
        self.work_in = _NO_WORK
        self.device_in = _NO_WORK

    def device_in_add(self, packet):
        self.device_in = self.append(packet, self.device_in)

    def work_in_add(self, packet):
        self.work_in = self.append(packet, self.work_in)


class _IdleTaskDataRecord(_RBObject):
    def __init__(self):
        self.control = 1
        self.count = 10_000


class _Packet(_RBObject):
    def __init__(self, link, identity, kind):
        self.link = link
        self.kind = kind
        self.identity = identity
        self.datum = 0
        self.data = [0] * _DATA_SIZE


class _TaskState(_RBObject):
    def __init__(self):
        self._task_holding = False
        self._task_waiting = False
        self._packet_pending = False

    def is_packet_pending(self):
        return self._packet_pending

    def is_task_waiting(self):
        return self._task_waiting

    def is_task_holding(self):
        return self._task_holding

    def set_task_holding(self, task):
        self._task_holding = task

    def set_task_waiting(self, task):
        self._task_waiting = task

    def packet_pending(self):
        self._packet_pending = True
        self._task_waiting = False
        self._task_holding = False
        return self

    def running(self):
        self._packet_pending = self._task_waiting = self._task_holding = False
        return self

    def waiting(self):
        self._packet_pending = self._task_holding = False
        self._task_waiting = True
        return self

    def waiting_with_packet(self):
        self._task_holding = False
        self._task_waiting = self._packet_pending = True
        return self

    def is_task_holding_or_waiting(self):
        return self._task_holding or (not self._packet_pending and self._task_waiting)

    def is_waiting_with_packet(self):
        return self._packet_pending and self._task_waiting and not self._task_holding

    @classmethod
    def with_running(cls):
        return cls().running()

    @classmethod
    def with_waiting(cls):
        return cls().waiting()

    @classmethod
    def with_waiting_with_packet(cls):
        return cls().waiting_with_packet()


class _TaskControlBlock(_TaskState):
    def __init__(
        self,
        link,
        identity,
        priority,
        initial_work_queue,
        initial_state,
        private_data,
        fn,
    ):
        super().__init__()
        self.link = link
        self.identity = identity
        self.function = fn
        self.priority = priority
        self._input = initial_work_queue
        self._handle = private_data

        self._packet_pending = initial_state.is_packet_pending()
        self._task_waiting = initial_state.is_task_waiting()
        self._task_holding = initial_state.is_task_holding()

    def add_input_and_check_priority(self, packet, old_task):
        if _NO_WORK == self._input:
            self._input = packet
            self._packet_pending = True
            if self.priority > old_task.priority:
                return self
        else:
            self._input = self.append(packet, self._input)

        return old_task

    def run_task(self):
        if self.is_waiting_with_packet():
            message = self._input
            self._input = message.link
            if _NO_WORK == self._input:
                self.running()
            else:
                self.packet_pending()
        else:
            message = _NO_WORK

        return self.function(message, self._handle)


class _WorkerTaskDataRecord(_RBObject):
    def __init__(self):
        self.destination = _HANDLER_A
        self.count = 0
